### PR TITLE
Data team

### PR DESCRIPTION
**Pull request by caracfinlay**
Sql script for database schema, including view for current KPI values and trigger for handling dates.

### CODE REVIEW

File: data-team/notebooks/schema.sql [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+CREATE extension if not exists moddatetime schema extensions;
+
+CREATE TABLE public.kpi_definition (
+    kpi_id int8 generated by default as identity,
+    kpi_name varchar(255) not null unique,
+    description varchar(255),
+    periodicity periodicity not null,
+    value_min int8,
+    value_max int8,
+    unit unit,
```
Comment (daczczcz1) [inline data-team/notebooks/schema.sql:34]: Perhaps it should be not null as well?

File: data-team/notebooks/schema.sql [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    created_at timestamp default current_timestamp,
```
Comment (daczczcz1) [inline data-team/notebooks/schema.sql:35]: Not null again?

File: data-team/notebooks/schema.sql [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+-- This will set the `updated_at` column on every update
+CREATE TRIGGER circle_updated_at before update on public.circle
+  for each row execute procedure moddatetime (updated_at);
+
+CREATE TABLE public.kpi_user (
+    user_id uuid not null default auth.uid (),
+    user_name character varying(255) not null unique,
+    role public.role not null,
+    created_at timestamp default current_timestamp,
```
Comment (daczczcz1) [inline data-team/notebooks/schema.sql:60]: Not null again?

File: data-team/notebooks/schema.sql [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+CREATE TABLE public.kpi_values_history (
+    kpi_value_history_id int8 generated by default as identity,
+    kpi_id int8 not null,
+    circle_id int8 not null,
+    user_id uuid not null default auth.uid (),
+    value numeric not null,
+    period_year int8 not null,
+    period_month int8 not null,
+    action action not null,
+    created_at timestamp default current_timestamp,
```
Comment (daczczcz1) [inline data-team/notebooks/schema.sql:128]: Not null again?

**Review by daczczcz1 — CHANGES_REQUESTED** [review]
Added some comments, overall it looks good - I would also rebase the branch on top of master and squash the commits (if not into just one, then maybe two: one for the SQL scripts another for the data, though it can be just one commit IMHO).

File: data-team/notebooks/schema.sql [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+      kpi_values_history.kpi_id,
+      kpi_values_history.circle_id,
+      kpi_values_history.period_year,
+      kpi_values_history.period_month
+  ) t2 ON t1.kpi_id = t2.kpi_id
+  AND t1.circle_id = t2.circle_id
+  AND t1.period_year = t2.period_year
+  AND t1.period_month = t2.period_month
+  AND t1.created_at = t2.updated_at;
+
```
Comment (Aurelieph) [inline data-team/notebooks/schema.sql:174]: new 2 views 
```suggestion
create view
  public.kpi_values_period_standardized as
select distinct
  on (
    kd.kpi_id,
    (
      case
        when kd.periodicity = 'yearly'::periodicity then date_trunc(
          'year'::text,
          t1.period_date::timestamp with time zone
        ) + '1 year'::interval - '1 day'::interval
        when kd.periodicity = 'monthly'::periodicity then date_trunc(
          'month'::text,
          t1.period_date::timestamp with time zone
        ) + '1 mon'::interval - '1 day'::interval
        when kd.periodicity = 'quarterly'::periodicity then date_trunc(
          'quarter'::text,
          t1.period_date::timestamp with time zone
        ) + '3 mons'::interval - '1 day'::interval
        when kd.periodicity = 'weekly'::periodicity then date_trunc(
          'week'::text,
          t1.period_date::timestamp with time zone
        ) + '7 days'::interval - '1 day'::interval
        else t1.period_date::timestamp with time zone
      end
    ),
    t1.circle_id
  ) t1.kpi_value_history_id,
  t1.kpi_id,
  t1.circle_id,
  t1.user_id,
  t1.value,
  t1.period_date,
  t1.action,
  t1.created_at,
  kd.periodicity as kpi_periodicity,
  case
    when kd.periodicity = 'yearly'::periodicity then date_trunc(
      'year'::text,
      t1.period_date::timestamp with time zone
    ) + '1 year'::interval - '1 day'::interval
    when kd.periodicity = 'monthly'::periodicity then date_trunc(
      'month'::text,
      t1.period_date::timestamp with time zone
    ) + '1 mon'::interval - '1 day'::interval
    when kd.periodicity = 'quarterly'::periodicity then date_trunc(
      'quarter'::text,
      t1.period_date::timestamp with time zone
    ) + '3 mons'::interval - '1 day'::interval
    when kd.periodicity = 'weekly'::periodicity then date_trunc(
      'week'::text,
      t1.period_date::timestamp with time zone
    ) + '7 days'::interval - '1 day'::interval
    else t1.period_date::timestamp with time zone
  end as standardized_date
from
  kpi_values_history t1
  join kpi_definition kd on t1.kpi_id = kd.kpi_id
order by
  kd.kpi_id,
  (
    case
      when kd.periodicity = 'yearly'::periodicity then date_trunc(
        'year'::text,
        t1.period_date::timestamp with time zone
      ) + '1 year'::interval - '1 day'::interval
      when kd.periodicity = 'monthly'::periodicity then date_trunc(
        'month'::text,
        t1.period_date::timestamp with time zone
      ) + '1 mon'::interval - '1 day'::interval
      when kd.periodicity = 'quarterly'::periodicity then date_trunc(
        'quarter'::text,
        t1.period_date::timestamp with time zone
      ) + '3 mons'::interval - '1 day'::interval
      when kd.periodicity = 'weekly'::periodicity then date_trunc(
        'week'::text,
        t1.period_date::timestamp with time zone
      ) + '7 days'::interval - '1 day'::interval
      else t1.period_date::timestamp with time zone
    end
  ) desc,
  t1.circle_id,
  t1.created_at desc;
  
  create view
  public.kpi_definition_with_latest_values as
select
  ckd.circle_kpidef_id,
  ckd.circle_id,
  kd.kpi_id,
  kd.kpi_name,
  kd.description,
  kd.periodicity,
  kd.value_min,
  kd.value_max,
  kd.unit,
  kd.created_at as kpi_created_at,
  kd.updated_at as kpi_updated_at,
  t1.latest_value,
  t1.latest_user_id,
  t1.latest_standardized_date
from
  circle_kpi_definition ckd
  join kpi_definition kd on ckd.kpi_id = kd.kpi_id
  left join (
    select
      kv.kpi_id,
      kv.circle_id,
      kv.standardized_date as latest_standardized_date,
      kv.value as latest_value,
      kv.user_id as latest_user_id
    from
      kpi_values_period_standardized kv
      join (
        select
          kpi_values_period_standardized.kpi_id,
          kpi_values_period_standardized.circle_id,
          max(kpi_values_period_standardized.standardized_date) as max_standardized_date
        from
          kpi_values_period_standardized
        group by
          kpi_values_period_standardized.kpi_id,
          kpi_values_period_standardized.circle_id
      ) max_dates on kv.kpi_id = max_dates.kpi_id
      and kv.circle_id = max_dates.circle_id
      and kv.standardized_date = max_dates.max_standardized_date
  ) t1 on ckd.kpi_id = t1.kpi_id
  and ckd.circle_id = t1.circle_id;
```

**Pull request closed without merging**